demos/http/http_demo_mutual_tls/http_demo_mutual_tls.c

```diff
@@ -0,0 +1,923 @@
+        returnStatus = EXIT_SUCCESS;
+    }
+
+    return returnStatus;
+}
+
+/*-----------------------------------------------------------*/
+
+static int32_t transportSend( HTTPNetworkContext_t * pNetworkContext,
+                              const void * pBuffer,
+                              size_t bytesToSend )
+{
+    int32_t bytesSent = 0;
+    int pollStatus = 0;
+    struct pollfd fileDescriptor;
+
+    /* Initialize the file descriptor. */
+    fileDescriptor.events = POLLOUT;
+    fileDescriptor.revents = 0;
+    /* Set the file descriptor for poll. */
+    fileDescriptor.fd = pNetworkContext->tcpSocket;
+
+    /* Poll the file descriptor to check if SSL_Write can be done now. */
+    pollStatus = poll( &fileDescriptor, 1, TRANSPORT_SEND_RECV_TIMEOUT_MS );
+
+    if( pollStatus > 0 )
+    {
+        bytesSent = ( int32_t ) SSL_write( pNetworkContext->pSslContext, pBuffer, bytesToSend );
+    }
+    else if( pollStatus == 0 )
+    {
+        LogDebug( ( "Timed out while polling SSL socket for write buffer availability." ) );
+    }
+    else
+    {
+        LogError( ( "Polling of the SSL socket for write buffer availability failed:"
+                    " status=%d",
+                    pollStatus ) );
+        bytesSent = -1;
+    }
+
+    return bytesSent;
+}
+
+/*-----------------------------------------------------------*/
+
+static int32_t transportRecv( HTTPNetworkContext_t * pNetworkContext,
+                              void * pBuffer,
+                              size_t bytesToRecv )
+{
+    int32_t bytesReceived = 0;
+    int pollStatus = -1, bytesAvailableToRead = 0;
+    struct pollfd fileDescriptor;
+
+    /* Initialize the file descriptor. */
+    fileDescriptor.events = POLLIN | POLLPRI;
+    fileDescriptor.revents = 0;
+    /* Set the file descriptor for poll. */
+    fileDescriptor.fd = SSL_get_fd( pNetworkContext->pSslContext );
+
+    /* Check if there are any pending data available for read. */
+    bytesAvailableToRead = SSL_pending( pNetworkContext->pSslContext );
+
+    /* Poll only if there is no data available yet to read. */
+    if( bytesAvailableToRead <= 0 )
+    {
+        pollStatus = poll( &fileDescriptor, 1, TRANSPORT_SEND_RECV_TIMEOUT_MS );
+    }
+
+    /* bytesAvailableToRead > 0 means that there was pending data to be read.
+     * pollStatus > 0 means that there was no pending data, but it became available
+     * during polling. If either holds true, read the available data. */
+    if( ( bytesAvailableToRead > 0 ) || ( pollStatus > 0 ) )
+    {
+        bytesReceived = ( int32_t ) SSL_read( pNetworkContext->pSslContext, pBuffer, bytesToRecv );
+    }
+    /* Poll timed out. */
+    else if( pollStatus == 0 )
+    {
+        LogInfo( ( "Poll timed out and there is no data to read from the buffer." ) );
+    }
+    else
+    {
+        LogError( ( "Poll returned with status = %d.", pollStatus ) );
+        bytesReceived = -1;
+    }
+
+    return bytesReceived;
+}
```

Comment:
Everything above this line to do with the network & TLS should be in a separate file since it has nothing to with HTTP.  They have opened a file called `http_demo_mutual_tls.c` to learn how to create and use an HTTP/TLS connection, but more than 75% of the file has nothing to do with HTTP.

Reply:
I completely agree. We already have a story for this that will put all TCP/TLS-related functions in the `platform/posix` directory. We will then refactor all demos for both MQTT and HTTP.